src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,1023 @@
+use std::{collections::BTreeMap, marker::PhantomData};
+
+use super::{EccInstructions, FixedPoints};
+use crate::constants::{self, FixedBase, Name};
+use ff::PrimeField;
+use halo2::{
+    arithmetic::{CurveAffine, FieldExt},
+    circuit::{CellValue, Chip, Config, Layouter, Loaded},
+    plonk::{Advice, Column, ConstraintSystem, Error, Fixed, Permutation, Selector},
+    poly::Rotation,
+};
+
+mod add;
+mod add_complete;
+mod double;
+mod mul;
+mod mul_fixed;
+mod mul_fixed_short;
+mod util;
+mod witness_point;
+mod witness_scalar_fixed;
+mod witness_scalar_fixed_short;
+mod witness_scalar_var;
```

Comment:
We don't need to separately decompose `ivk` for variable-base scalar mul; we can merge the decomposition into the scalar mul.